src/compose/ComposeBox.js

```diff
@@ -153,270 +136,352 @@ const updateTextInput = (textInput, text) => {
-    this.inputBlurTimeoutId = null;
-  }
+  useEffect(
+    () => () => {
+      clearTimeout(inputBlurTimeoutId.current);
+      inputBlurTimeoutId.current = null;
+    },
+    [],
+  );
+
+  const prevMessageInputState = usePrevious(messageInputState);
+  useEffect(() => {
+    const { dispatch, isEditing, narrow } = props;
+    const messageInputValue = messageInputState.value;
+    const prevMessageInputValue = prevMessageInputState?.value;
+
+    if (prevMessageInputValue !== messageInputValue) {
+      if (messageInputValue.length === 0) {
+        dispatch(sendTypingStop(narrow));
+      } else {
+        dispatch(sendTypingStart(narrow));
+      }
+      if (!isEditing) {
+        dispatch(draftUpdate(narrow, messageInputValue));
+      }
+    }
+  }, [props, messageInputState, prevMessageInputState]);
 
-  updateIsFocused = () => {
-    this.setState(state => ({
-      ...state,
-      isFocused: state.isMessageFocused || state.isTopicFocused,
-    }));
-  };
+  const updateIsFocused = useCallback(() => {
+    setFocusState(s => ({ ...s, either: s.message || s.topic }));
+  }, [setFocusState]);
 
-  getCanSelectTopic = () => {
-    const { isEditing, narrow } = this.props;
+  const getCanSelectTopic = useCallback(() => {
+    const { isEditing, narrow } = props;
     if (isEditing) {
       return isStreamOrTopicNarrow(narrow);
     }
     if (!isStreamNarrow(narrow)) {
       return false;
     }
-    return this.state.isFocused;
-  };
-
-  setMessageInputValue = (message: string) => {
-    updateTextInput(this.messageInputRef.current, message);
-    this.handleMessageChange(message);
-  };
-
-  setTopicInputValue = (topic: string) => {
-    updateTextInput(this.topicInputRef.current, topic);
-    this.handleTopicChange(topic);
-  };
-
-  insertMessageTextAtCursorPosition = (text: string) => {
-    const { message, selection } = this.state;
+    return focusState.either;
+  }, [props, focusState.either]);
 
-    this.setMessageInputValue(
-      message.slice(0, selection.start) + text + message.slice(selection.end, message.length),
-    );
-  };
+  const handleMessageChange = useCallback(
+    (newValue: string) => {
+      setMessageInputState(s => ({ ...s, value: newValue }));
+      setIsMenuExpanded(false);
+    },
+    [setMessageInputState, setIsMenuExpanded],
+  );
+
+  const setMessageInputValue = useCallback(
+    (updater: (typeof messageInputState) => string) => {
+      updateTextInput(messageInputRef.current, updater(messageInputState));
+      setMessageInputState(s => ({ ...s, value: updater(s) }));
+      setIsMenuExpanded(false);
+    },
+    [messageInputState, setIsMenuExpanded],
+  );
 
-  insertVideoCallLinkAtCursorPosition = (url: string) => {
-    const { _ } = this.props;
-    const linkMessage = _('Click to join video call');
-    const linkText = `[${linkMessage}](${url})`;
+  const handleTopicChange = useCallback(
+    (newValue: string) => {
+      setTopicInputState({ value: newValue });
```

Comment:
nit:
```suggestion
    (value: string) => {
      setTopicInputState({ value });
```

Or perhaps better: `topic`. But `newValue` doesn't really add any information beyond `value`.